src/libraries/Microsoft.PowerFx.Core/Public/CheckResult.cs

```diff
@@ -17,115 +19,413 @@
+        internal TexlBinding ApplyBindingInternal()
         {
+            if (_binding == null)
+            {
+                if (!this._setBindingCalled)
+                {
+                    throw new InvalidOperationException($"Must call {nameof(SetBindingInfo)} before calling {nameof(ApplyBinding)}.");
+                }
+
+                (var binding, var combinedSymbols) = Source.ComputeBinding(this);
+
+                this.ThrowIfSymbolsChanged();
+                
+                // Don't modify any fields until after we've verified the symbols haven't change.
+
+                this._binding = binding;
+                this.AllSymbols = combinedSymbols;
+
+                // Add the errors
+                var bindingErrors = ExpressionError.New(binding.ErrorContainer.GetErrors(), CultureInfo);
+                _errors.AddRange(bindingErrors);
+
+                if (this.IsSuccess && !this.HasDeferredArgsWarning)
+                {
+                    // TODO: Fix FormulaType.Build to not throw exceptions for Enum types then remove this check
+                    if (binding.ResultType.Kind != DKind.Enum)
+                    {
+                        this.ReturnType = FormulaType.Build(binding.ResultType);
+                    }
+                }
+            }
+
+            return _binding;
         }
 
-        internal CheckResult(ParseResult parse, CultureInfo locale, TexlBinding binding = null)
+        /// <summary>
+        /// Compute the dependencies. Called after binding. 
+        /// </summary>
+        public void ApplyDependencyAnalysis()
         {
-            Parse = parse ?? throw new ArgumentNullException(nameof(parse));
-            CultureInfo = locale;
+            var binding = this.Binding; // will throw if binding wasn't run
+            this._topLevelIdentifiers = DependencyFinder.FindDependencies(binding.Top, binding);
+        }
 
-            _binding = binding;
+        // Flag to ensure Post Checks are only invoked once. 
+        private bool _invokingPostCheck;
 
-            Errors = Parse.Errors.Concat(BindingErrors);
+        /// <summary>
+        /// Calculate all errors. 
+        /// Invoke Binding and any engine-specific errors via <see cref="Engine.PostCheck(CheckResult)"/>. 
+        /// </summary>
+        /// <returns></returns>
+        public IEnumerable<ExpressionError> ApplyErrors()
+        {
+            if (!_invokingPostCheck)
+            {
+                _invokingPostCheck = true;
+
+                // Errors require Binding, Parse 
+                var binding = ApplyBindingInternal();
+
+                // Plus engine's may have additional constaints. 
+                // PostCheck may refer to binding. 
+                var extraErrors = Source.InvokePostCheck(this);
+
+                _errors.AddRange(extraErrors);
+            }
+            
+            return this.Errors;
         }
 
-        public void ThrowOnErrors()
+        internal IRResult ApplyIR()

```

Comment:
>ApplyIR [](http://example.com/codeflow?start=26&length=7)

GetIR() ?

Reply:
The naming pattern here is Apply*() for all these operations. (ApplyParse,ApplyBinding,ApplyDependency,etc)
It's a method , not a property, because it does significant explicit work (and can be side-effecting).